### PR TITLE
Add stable diffusion onnx runtime pipeline

### PR DESCRIPTION
**Pull request by echarlaix**
In this PR we add a stable diffusion pipeline enabling the ONNX export of stable diffusion models (`export` must be set to `True`) and ONNX Runtime inference.

```python
from optimum.onnxruntime import ORTStableDiffusionPipeline

model_id = "runwayml/stable-diffusion-v1-5"
stable_diffusion = ORTStableDiffusionPipeline.from_pretrained(model_id , export=True)
prompt = "sailing ship in storm by Leonardo da Vinci"
images = stable_diffusion(prompt).images
```

### CODE REVIEW

**Review by regisss — COMMENTED** [review]
Awesome @echarlaix!!
I left a lot of minor comments. Also, the doc you wrote is not in the `_toctree.yml` so it does not appear in this PR doc.

File: docs/source/onnxruntime/usage_guides/models.mdx [hunk 1]
```diff
@@ -91,3 +91,24 @@ Here is an example of how you can load a T5 model to the ONNX format and run inf
...
+- The U-NET component
+- The VAE decoder component
+
+Here is an example of how you can load a stable diffusion model, export it to the ONNX format and run inference using ONNX Runtime:
+
+```python
+from optimum.onnxruntime import ORTStableDiffusionPipeline
+
+model_id = "runwayml/stable-diffusion-v1-5"
+stable_diffusion = ORTStableDiffusionPipeline.from_pretrained(model_id, export=True)
```
Comment (fxmarty) [inline docs/source/onnxruntime/usage_guides/models.mdx:109]: It's a bit tricky to advertise `export=True` given that the cache has been disabled for now and the export for somewhat large models like stable diffusion can be quite slow. So I'd advocate either to add a section about offline loading, or add back a cache in a following PR.

File: optimum/onnxruntime/__init__.py [hunk 2]
```diff
@@ -57,6 +59,14 @@
     ],
 }
 
+try:
+    if not is_diffusers_available():
+        raise OptionalDependencyNotAvailable()
+except OptionalDependencyNotAvailable:
+    pass
+else:
+    _import_structure["modeling_diffusion"] = ["ORTStableDiffusionPipeline"]
```
Comment (fxmarty) [inline optimum/onnxruntime/__init__.py:68]: would
```suggestion
try:
    if not is_diffusers_available():
        raise OptionalDependencyNotAvailable()
    _import_structure["modeling_diffusion"] = ["ORTStableDiffusionPipeline"]
except OptionalDependencyNotAvailable:
    pass
```

be different?

File: optimum/onnxruntime/__init__.py [hunk 3]
```diff
@@ -89,12 +99,15 @@
         ONNX_WEIGHTS_NAME,
         ORTQuantizableOperator,
     )
+
+    try:
```
Comment (fxmarty) [inline optimum/onnxruntime/__init__.py:103]: same question

File: optimum/onnxruntime/modeling_diffusion.py [hunk 4]
```diff
@@ -0,0 +1,412 @@
...
+            provider=provider,
+            session_options=session_options,
+            provider_options=provider_options,
+        )
+
+        if model_save_dir is None:
+            model_save_dir = new_model_save_dir
+
+        if use_io_binding:
+            logger.warning("OBinding is not yet available, `use_io_binding` set to False.")
```
Comment (fxmarty) [inline optimum/onnxruntime/modeling_diffusion.py:273]: I would rather raise an error here.

File: optimum/onnxruntime/modeling_diffusion.py [hunk 5]
```diff
@@ -0,0 +1,412 @@
...
+            save_dir_path,
+            config=config,
+            provider=provider,
+            session_options=session_options,
+            provider_options=provider_options,
+            use_io_binding=use_io_binding,
+            model_save_dir=save_dir,
+        )
+
+    def to(self, device: Union[torch.device, str, int]):
```
Comment (fxmarty) [inline optimum/onnxruntime/modeling_diffusion.py:349]: Given that this code is duplicate in `modeling_decoder.py`, `modeling_seq2seq.py` and `modeling_diffusion.py`, I think it can make sense to abstract it. But I think it can be done in an other PR.

File: optimum/onnxruntime/modeling_diffusion.py [hunk 6]
```diff
@@ -0,0 +1,412 @@
...
+    @classmethod
+    def _load_config(cls, config_name_or_path: Union[str, os.PathLike], **kwargs):
+        return cls.load_config(config_name_or_path, **kwargs)
+
+    def _save_config(self, save_directory):
+        self.save_config(save_directory)
+
+
+class ORTModelTextEncoder(ORTModelPart):
+    def forward(self, input_ids: np.ndarray):
```
Comment (fxmarty) [inline optimum/onnxruntime/modeling_diffusion.py:383]: torch.Tensor are not accepted?

File: optimum/onnxruntime/modeling_diffusion.py [hunk 7]
```diff
@@ -0,0 +1,412 @@
...
+        outputs = self.session.run(None, onnx_inputs)
+        return outputs
+
+
+class ORTModelUnet(ORTModelPart):
+    def __init__(self, session: ort.InferenceSession, parent_model: "ORTModel"):
+        super().__init__(session, parent_model)
+        self.input_dtype = {inputs.name: ORT_TO_NP_TYPE[inputs.type] for inputs in self.session.get_inputs()}
+
+    def forward(self, sample: np.ndarray, timestep: np.ndarray, encoder_hidden_states: np.ndarray):
```
Comment (fxmarty) [inline optimum/onnxruntime/modeling_diffusion.py:396]: same question

File: optimum/onnxruntime/modeling_diffusion.py [hunk 8]
```diff
@@ -0,0 +1,412 @@
...
+            "sample": sample,
+            "timestep": timestep,
+            "encoder_hidden_states": encoder_hidden_states,
+        }
+        outputs = self.session.run(None, onnx_inputs)
+        return outputs
+
+
+class ORTModelVaeDecoder(ORTModelPart):
+    def forward(self, latent_sample: np.ndarray):
```
Comment (fxmarty) [inline optimum/onnxruntime/modeling_diffusion.py:407]: same

File: optimum/onnxruntime/modeling_diffusion.py [hunk 9]
```diff
@@ -0,0 +1,412 @@
...
+            "input_ids": input_ids,
+        }
+        outputs = self.session.run(None, onnx_inputs)
+        return outputs
+
+
+class ORTModelUnet(ORTModelPart):
+    def __init__(self, session: ort.InferenceSession, parent_model: "ORTModel"):
+        super().__init__(session, parent_model)
+        self.input_dtype = {inputs.name: ORT_TO_NP_TYPE[inputs.type] for inputs in self.session.get_inputs()}
```
Comment (fxmarty) [inline optimum/onnxruntime/modeling_diffusion.py:394]: Why is this needed?

File: tests/onnxruntime/test_modeling.py [hunk 10]
```diff
@@ -191,7 +200,9 @@ def _setup(self, model_args: Dict):
         if "use_cache" in model_args and model_args["use_cache"] is True:
             task = task + "-with-past"
 
-        if task not in TasksManager.get_supported_tasks_for_model_type(model_arch.replace("_", "-"), exporter="onnx"):
+        if "use_cache" in model_args and task not in TasksManager.get_supported_tasks_for_model_type(
```
Comment (fxmarty) [inline tests/onnxruntime/test_modeling.py:203]: thanks!

File: tests/onnxruntime/test_modeling.py [hunk 11]
```diff
@@ -3281,3 +3561,122 @@ def test_find_untested_architectures(self, task: str, test_class):
...
+
+    @parameterized.expand(
+        grid_parameters(
+            {"model_arch": SUPPORTED_ARCHITECTURES, "provider": ["CUDAExecutionProvider", "TensorrtExecutionProvider"]}
+        )
+    )
+    @require_torch_gpu
+    @pytest.mark.gpu_test
+    @require_diffusers
+    def test_pipeline_on_gpu(self, test_name: str, model_arch: str, provider: str):
```
Comment (fxmarty) [inline tests/onnxruntime/test_modeling.py:3673]: Could you make sure the GPU tests pass?

**Review by fxmarty — APPROVED** [review]
Thank you, looks great!

File: optimum/onnxruntime/__init__.py (hunk 2 above)
Comment (michaelbenayoun) [inline optimum/onnxruntime/__init__.py:68]: What's the goal with this?

File: optimum/onnxruntime/base.py [hunk 12]
```diff
@@ -41,9 +42,11 @@ def __init__(
...
         self.session = session
         self.parent_model = parent_model
-        self.normalized_config = NormalizedConfigManager.get_normalized_config_class(
-            self.parent_model.config.model_type
-        )(self.parent_model.config)
+        self.normalized_config = None
+        if isinstance(self.parent_model.config, PretrainedConfig):
+            self.normalized_config = NormalizedConfigManager.get_normalized_config_class(
+                self.parent_model.config.model_type
+            )(self.parent_model.config)
```
Comment (michaelbenayoun) [inline optimum/onnxruntime/base.py:49]: I do not like `self.normalized_config` possibly being `None` because we would not expect that. I get that this is for Stable Diffusion. Could we make a subclass specifically for StableDiffusion?

File: optimum/onnxruntime/modeling_diffusion.py [hunk 13]
```diff
@@ -0,0 +1,412 @@
...
+                A config dictionary from which the model components will be instantiated. Make sure to only load
+                configuration files of compatible classes.
+            feature_extractor (`Optional[CLIPFeatureExtractor]`, defaults to `None`):
+                A model extracting features from generated images to be used as inputs for the `safety_checker`
+            use_io_binding (`Optional[bool]`, defaults to `None`):
+                Whether use IOBinding during inference to avoid memory copy between the host and devices. Defaults to
+                `True` if the device is CUDA, otherwise defaults to `False`.
+            model_save_dir (`Optional[str]`, defaults to `None`):
+                The directory under which the model exported to ONNX was saved.
+        """
```
Comment (michaelbenayoun) [inline optimum/onnxruntime/modeling_diffusion.py:88]: Handle the `self.normalized_config = None` thing here.

File: optimum/onnxruntime/modeling_diffusion.py [hunk 14]
```diff
@@ -0,0 +1,412 @@
...
+                The directory where to save the model files.
+            vae_decoder_file_name (`str`, defaults to `optimum.onnxruntime.utils.ONNX_WEIGHTS_NAME`):
+                The VAE decoder model file name. Overwrites the default file name and allows one to save the VAE decoder model
+                with a different name.
+            text_encoder_file_name (`str`, defaults to `optimum.onnxruntime.utils.ONNX_WEIGHTS_NAME`):
+                The text encoder model file name. Overwrites the default file name and allows one to save the text encoder model
+                with a different name.
+            unet_file_name (`str`, defaults to `optimum.onnxruntime.ONNX_WEIGHTS_NAME`):
+                The U-NET model file name. Overwrites the default file name and allows one to save the U-NET model
+                with a different name.
```
Comment (michaelbenayoun) [inline optimum/onnxruntime/modeling_diffusion.py:173]: I would avoid adding a docstring for protected methods, except if this is absolutely needed.
This becomes hard to maintain along time, and usually it does not bring much value.

File: optimum/onnxruntime/modeling_diffusion.py [hunk 15]
```diff
@@ -0,0 +1,412 @@
...
+
+    def __call__(self, *args, **kwargs):
+        return StableDiffusionPipelineMixin.__call__(self, *args, **kwargs)
+
+    @classmethod
+    def _load_config(cls, config_name_or_path: Union[str, os.PathLike], **kwargs):
+        return cls.load_config(config_name_or_path, **kwargs)
+
+    def _save_config(self, save_directory):
+        self.save_config(save_directory)
```
Comment (michaelbenayoun) [inline optimum/onnxruntime/modeling_diffusion.py:379]: Same comment

File: optimum/pipelines/diffusers/pipeline_utils.py [hunk 16]
```diff
@@ -0,0 +1,53 @@
...
+from PIL import Image
+from tqdm.auto import tqdm
+
+
+class DiffusionPipelineMixin(ConfigMixin):
+    # Copied from https://github.com/huggingface/diffusers/blob/v0.12.1/src/diffusers/pipelines/pipeline_utils.py#L812
+    @staticmethod
+    def numpy_to_pil(images):
+        """
+        Convert a numpy image or a batch of images to a PIL image.
```
Comment (michaelbenayoun) [inline optimum/pipelines/diffusers/pipeline_utils.py:26]:
```suggestion
        Converts a numpy image or a batch of images to a PIL image.
```

File: optimum/utils/constant.py [hunk 17]
```diff
@@ -0,0 +1,20 @@
+# Copyright 2022 The HuggingFace Inc. team. All rights reserved.
```
Comment (michaelbenayoun) [inline optimum/utils/constant.py:1]:
```suggestion
# Copyright 2023 The HuggingFace Inc. team. All rights reserved.
```
Also this might be the case in other files as well!

File: optimum/utils/doc.py [hunk 18]
```diff
@@ -1,3 +1,18 @@
+# Copyright 2022 The HuggingFace Inc. team. All rights reserved.
```
Comment (michaelbenayoun) [inline optimum/utils/doc.py:1]: Same comment

File: optimum/utils/modeling_utils.py [hunk 19]
```diff
@@ -1,3 +1,18 @@
+# Copyright 2022 The HuggingFace Inc. team. All rights reserved.
```
Comment (michaelbenayoun) [inline optimum/utils/modeling_utils.py:1]: Same comment

File: optimum/modeling_base.py [hunk 20]
```diff
@@ -136,6 +137,13 @@ def _save_pretrained(self, save_directory):
         """
         raise NotImplementedError
 
+    def _save_config(self, save_directory):
+        """
+        Saves a model configuration into a directory, so that it can be re-loaded using the
+        [`from_pretrained`] class method.
+        """
+        self.config.save_pretrained(save_directory)
```
[the remark this answers is not in the document]
Comment (michaelbenayoun) [reply optimum/modeling_base.py:145]: Alright!

**Comment by echarlaix** [comment]
Thanks for the review @regisss !! The documentation of `ORTStableDiffusionPipeline` was added in onnxruntime/usage_guides/models.mdx, which seems to appear on the generated [documentation](https://moon-ci-docs.huggingface.co/docs/optimum/pr_786/en/onnxruntime/usage_guides/models#export-and-inference-of-stable-diffusion-models)

**Comment by regisss** [comment]
> Thanks for the review @regisss !! The documentation of `ORTStableDiffusionPipeline` was added in onnxruntime/usage_guides/models.mdx, which seems to appear on the generated [documentation](https://moon-ci-docs.huggingface.co/docs/optimum/pr_786/en/onnxruntime/usage_guides/models#export-and-inference-of-stable-diffusion-models)

Geez my bad :grin: I don't know why I thought it was a new file, I'll take a look at it!

**Comment by fxmarty** [comment]
@echarlaix Could you trigger manually these CI on main after merging this PR?

https://github.com/huggingface/optimum/actions/workflows/test_onnxruntime_gpu.yml
https://github.com/huggingface/optimum/actions/workflows/test_exporters_gpu.yml
https://github.com/huggingface/optimum/actions/workflows/test_exporters_slow.yml

I'd like to have them run on the latest commit before the release. Thanks!

**Comment by echarlaix** [comment]
> @echarlaix Could you trigger manually these CI on main after merging this PR?

Will do !

File: tests/onnxruntime/test_modeling.py (hunk 11 above)
Comment (echarlaix) [reply tests/onnxruntime/test_modeling.py:4046]: Just checked locally, all the newly added tests are passing !

File: optimum/onnxruntime/modeling_diffusion.py [hunk 21]
```diff
@@ -0,0 +1,423 @@
+import importlib
```
Comment (michaelbenayoun) [inline optimum/onnxruntime/modeling_diffusion.py:1]: Missing licence and module docstring.

File: optimum/onnxruntime/modeling_diffusion.py [hunk 22]
```diff
@@ -0,0 +1,423 @@
...
+        vae_decoder_session: ort.InferenceSession,
+        text_encoder_session: ort.InferenceSession,
+        unet_session: ort.InferenceSession,
+        tokenizer: CLIPTokenizer,
+        scheduler: Union[DDIMScheduler, PNDMScheduler, LMSDiscreteScheduler],
+        config: Dict[str, Any],
+        feature_extractor: Optional[CLIPFeatureExtractor] = None,
+        use_io_binding: Optional[bool] = None,
+        model_save_dir: Optional[Union[str, Path, TemporaryDirectory]] = None,
+        **kwargs,
```
Comment (michaelbenayoun) [inline optimum/onnxruntime/modeling_diffusion.py:63]: Why?

File: optimum/onnxruntime/modeling_diffusion.py [hunk 23]
```diff
@@ -0,0 +1,423 @@
...
+
+        return vae_decoder_session, text_encoder_session, unet_session
+
+    def _save_pretrained(
+        self,
+        save_directory: Union[str, Path],
+        vae_decoder_file_name: str = ONNX_WEIGHTS_NAME,
+        text_encoder_file_name: str = ONNX_WEIGHTS_NAME,
+        unet_file_name: str = ONNX_WEIGHTS_NAME,
+        **kwargs,
```
Comment (michaelbenayoun) [inline optimum/onnxruntime/modeling_diffusion.py:155]: Why?

File: optimum/onnxruntime/modeling_diffusion.py [hunk 24]
```diff
@@ -0,0 +1,423 @@
...
+        vae_decoder_file_name: str = ONNX_WEIGHTS_NAME,
+        text_encoder_file_name: str = ONNX_WEIGHTS_NAME,
+        unet_file_name: str = ONNX_WEIGHTS_NAME,
+        local_files_only: bool = False,
+        provider: str = "CPUExecutionProvider",
+        session_options: Optional[ort.SessionOptions] = None,
+        provider_options: Optional[Dict[str, Any]] = None,
+        use_io_binding: Optional[bool] = None,
+        model_save_dir: Optional[Union[str, Path, TemporaryDirectory]] = None,
+        **kwargs,
```
Comment (michaelbenayoun) [inline optimum/onnxruntime/modeling_diffusion.py:200]: Why?

File: optimum/onnxruntime/modeling_diffusion.py [hunk 25]
```diff
@@ -0,0 +1,423 @@
...
+            allow_patterns += list(
+                {
+                    vae_decoder_file_name,
+                    text_encoder_file_name,
+                    unet_file_name,
+                    SCHEDULER_CONFIG_NAME,
+                    CONFIG_NAME,
+                    cls.config_name,
+                }
+            )
```
Comment (michaelbenayoun) [inline optimum/onnxruntime/modeling_diffusion.py:221]: Doing set -> list because you want only unique values?

File: optimum/onnxruntime/modeling_diffusion.py [hunk 26]
```diff
@@ -0,0 +1,423 @@
...
+            use_auth_token=use_auth_token,
+            local_files_only=local_files_only,
+            force_download=force_download,
+            trust_remote_code=trust_remote_code,
+        )
+        output_names = [
+            os.path.join(DIFFUSION_MODEL_TEXT_ENCODER_SUBFOLDER, ONNX_WEIGHTS_NAME),
+            os.path.join(DIFFUSION_MODEL_UNET_SUBFOLDER, ONNX_WEIGHTS_NAME),
+            os.path.join(DIFFUSION_MODEL_VAE_ENCODER_SUBFOLDER, ONNX_WEIGHTS_NAME),
+            os.path.join(DIFFUSION_MODEL_VAE_DECODER_SUBFOLDER, ONNX_WEIGHTS_NAME),
```
Comment (michaelbenayoun) [inline optimum/onnxruntime/modeling_diffusion.py:310]: If those are paths, we can use the `/` operator.

File: optimum/onnxruntime/modeling_diffusion.py (hunk 22 above)
Comment (echarlaix) [reply optimum/onnxruntime/modeling_diffusion.py:63]: Removed

File: optimum/onnxruntime/modeling_diffusion.py (hunk 23 above)
Comment (echarlaix) [reply optimum/onnxruntime/modeling_diffusion.py:155]: Removed

File: optimum/onnxruntime/modeling_diffusion.py (hunk 24 above)
Comment (echarlaix) [reply optimum/onnxruntime/modeling_diffusion.py:200]: Argument `trust_remote_code` and `force_download` are not used (should be corrected in `OptimizedModel.from_pretrained` in an other PR)

File: optimum/onnxruntime/modeling_diffusion.py (hunk 25 above)
Comment (echarlaix) [reply optimum/onnxruntime/modeling_diffusion.py:221]: Yes

File: optimum/onnxruntime/modeling_diffusion.py (hunk 21 above)
Comment (echarlaix) [reply optimum/onnxruntime/modeling_diffusion.py:1]: Added

File: optimum/onnxruntime/modeling_diffusion.py (hunk 24 above)
Comment (michaelbenayoun) [reply optimum/onnxruntime/modeling_diffusion.py:212]: Then add explicitly `trsut_remote_code` and `force_download`, using `kwargs` when absolutely not needed is quite bug-prone.
